handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -236,56 +236,66 @@ protected void deallocate() {
         super(peerHost, peerPort);
         OpenSsl.ensureAvailability();
         leak = leakDetection ? leakDetector.track(this) : null;
-        this.alloc = checkNotNull(alloc, "alloc");
-        apn = (OpenSslApplicationProtocolNegotiator) context.applicationProtocolNegotiator();
-        session = new OpenSslSession(context.sessionContext());
-        clientMode = context.isClient();
-        engineMap = context.engineMap;
-        rejectRemoteInitiatedRenegotiation = context.getRejectRemoteInitiatedRenegotiation();
-        localCerts = context.keyCertChain;
-        keyMaterialManager = context.keyMaterialManager();
-        enableOcsp = context.enableOcsp;
-        this.jdkCompatibilityMode = jdkCompatibilityMode;
-        Lock readerLock = context.ctxLock.readLock();
-        readerLock.lock();
-        final long finalSsl;
+        boolean success = false;
         try {
-            finalSsl = SSL.newSSL(context.ctx, !context.isClient());
-        } finally {
-            readerLock.unlock();
-        }
-        synchronized (this) {
-            ssl = finalSsl;
+            clientMode = context.isClient();
+            engineMap = context.engineMap;
+
+            this.alloc = checkNotNull(alloc, "alloc");
+            apn = (OpenSslApplicationProtocolNegotiator) context.applicationProtocolNegotiator();
+            session = new OpenSslSession(context.sessionContext());
+            rejectRemoteInitiatedRenegotiation = context.getRejectRemoteInitiatedRenegotiation();
+            localCerts = context.keyCertChain;
+            keyMaterialManager = context.keyMaterialManager();
+            enableOcsp = context.enableOcsp;
+            this.jdkCompatibilityMode = jdkCompatibilityMode;
+            Lock readerLock = context.ctxLock.readLock();
+            readerLock.lock();
+            final long finalSsl;
             try {
-                networkBIO = SSL.bioNewByteBuffer(ssl, context.getBioNonApplicationBufferSize());
+                finalSsl = SSL.newSSL(context.ctx, !context.isClient());
+            } finally {
+                readerLock.unlock();
+            }
+            synchronized (this) {
```

Comment:
I think it would be safer if you just assigned `refCnt` here, after you have created `ssl`.   That would fix the problem of exceptions being thrown in initializers, or in super(), or in OpenSsl.ensureAvailability();